src/remotedevice.cpp

```diff
@@ -0,0 +1,188 @@
+        } else {
+            m_sshBinary = binary;
+        }
+    };
+
+    auto settings = Settings::instance();
+    findSSHBinary(settings->sshPath());
+    connect(settings, &Settings::sshPathChanged, this, findSSHBinary);
+}
+
+RemoteDevice::~RemoteDevice() = default;
+
+void RemoteDevice::connectToDevice(const QString& device)
+{
+    if (m_connection) {
+        disconnect();
+    }
+    auto config = KSharedConfig::openConfig()->group("SSH");
+    if (!config.hasGroup(device) && config.group(device).exists()) {
+        emit failedToConnect();
+        return;
+    }
+    m_config = config.group(device);
+
+    m_connection = sshProcess({});
+
+    connect(m_connection, qOverload<int, QProcess::ExitStatus>(&QProcess::finished), this,
+            [connection = m_connection, path = m_tempDir.path(), this](int exitCode, QProcess::ExitStatus) {
+                if (exitCode != 0) {
+                    emit failedToConnect();
+                } else {
+                    emit disconnected();
+                }
+                qDebug() << exitCode << connection->readAll();
+                connection->deleteLater();
+            });
+
+    m_connection->start();
+}
+
+void RemoteDevice::disconnect()
+{
+    if (m_connection) {
+        if (m_connection->state() == QProcess::ProcessState::Running) {
+            // ssh stops once you clode the write channel
+            // we then use the finished signal for cleanup
+            m_connection->closeWriteChannel();
+        }
+
+        m_connection = nullptr;
+    }
+}
+
+bool RemoteDevice::isConnected() const
+{
+    return QFile(m_tempDir.path() + QStringLiteral("/ssh")).exists();
+}
+
+bool RemoteDevice::checkIfProgramExists(const QString& program) const
+{
+    auto ssh = sshProcess({QStringLiteral("command"), program});
+    ssh->start();
+    ssh->waitForFinished();
+    auto exitCode = ssh->exitCode();
+    ssh->deleteLater();
+    // 128 -> not found
+    // perf will return 1 and display the help message
+    return exitCode != 128;
+}
+
+bool RemoteDevice::checkIfDirectoryExists(const QString& directory) const
+{
+    auto ssh = sshProcess({QStringLiteral("test"), QStringLiteral("-d"), directory});
+    ssh->start();
+    ssh->waitForFinished();
+    auto exitCode = ssh->exitCode();
+    ssh->deleteLater();
+    return exitCode == 0;
+}
+
+bool RemoteDevice::checkIfFileExists(const QString& file) const
+{
+    auto ssh = sshProcess({QStringLiteral("test"), QStringLiteral("-f"), file});
+    ssh->start();
+    ssh->waitForFinished();
+    auto exitCode = ssh->exitCode();
+    ssh->deleteLater();
+    return exitCode == 0;
+}
+
+QByteArray RemoteDevice::getProgramOutput(const QStringList& args) const
+{
+    auto ssh = sshProcess(args);
+    ssh->start();
+    ssh->waitForFinished();
+    auto output = ssh->readAllStandardOutput();
+    ssh->deleteLater();
+    return output;
+}
+
+QProcess* RemoteDevice::sshProcess(const QStringList& args)
```

Comment:
I don't get it, why do we have a const and a non-const version of this?

Reply:
make it so that the `const` version is enough, i.e. don't rely on QObject ownership, just return a `std::unique_ptr<QProcess>`.